File: javascript/src/problems101_200/lru-cache.js
export default class LruCache {
  capacity = 0;
  cache = new Map();

  constructor(capacity) {
    this.capacity = capacity;
  }

  get(key) {
    if (!this.cache.has(key)) {
      return -1;
    }
    const value = this.cache.get(key);
    this.cache.delete(key);
    this.cache.set(key, value); // put on end
    return value;
  }

  put(key, value) {
    console.log(this.cache.size + ' ' + this.capacity);
    if (this.cache.has(key)) {
      this.cache.delete(key);
    } else if (this.cache.size === this.capacity) {
      this.cache.delete(this.cache.keys().next().value);
    }
    this.cache.set(key, value);
  }
}

File: python/src/problems101_200/lru_cache.py
class LruCache:
    def __init__(self, capacity: int):
        self.capacity: int = capacity
        self.cache: dict[int, int] = {}

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        value: int = self.cache.pop(key)
        self.cache[key] = value  # put on end
        return value

    def set(self, key: int, value: int):
        if key in self.cache:
            self.cache.pop(key)
        elif len(self.cache) == self.capacity:
            self.cache.pop(next(iter(self.cache)))
        self.cache[key] = value
